prover/src/constraints/evaluator/lagrange.rs

```diff
@@ -0,0 +1,289 @@
+    /// That is, returns a vector of the form `[1 / div_0, ..., 1 / div_n]`, where `div_i` is the
+    /// divisor for the Lagrange kernel boundary constraint at the i'th row of the constraint
+    /// evaluation domain.
+    fn compute_boundary_divisors_inv(&self, domain: &StarkDomain<E::BaseField>) -> Vec<E> {
+        let mut boundary_denominator_evals = Vec::with_capacity(domain.ce_domain_size());
+        for step in 0..domain.ce_domain_size() {
+            let domain_point = domain.get_ce_x_at(step);
+            let boundary_denominator = self
+                .lagrange_kernel_constraints
+                .boundary
+                .evaluate_denominator_at(domain_point.into());
+            boundary_denominator_evals.push(boundary_denominator);
+        }
+
+        batch_inversion(&boundary_denominator_evals)
+    }
+}
+
+/// Holds all the transition constraint inverse divisor evaluations over the constraint evaluation
+/// domain.
+///
+/// [`LagrangeKernelTransitionConstraintsDivisor`] takes advantage of some structure in the
+/// divisors' evaluations. Recall that the divisor for the i'th transition constraint is `x^(2^i) -
+/// 1`. When substituting `x` for each value of the constraint evaluation domain, for constraints
+/// `i>0`, the divisor evaluations "wrap-around" such that some values repeat. For example,
+///
+/// i=0: no repetitions
+/// i=1: the first half of the buffer is equal to the second half
+/// i=2: each 1/4th of the buffer are equal
+/// i=3: each 1/8th of the buffer are equal
+/// ...
+/// Therefore, we only compute the non-repeating section of the buffer in each iteration, and index
+/// into it accordingly.
+struct LagrangeKernelTransitionConstraintsDivisor<E: FieldElement> {
+    divisor_evals_inv: Vec<E>,
+
+    // Precompute the indices into `divisors_evals_inv` of the slices that correspond to each
+    // transition constraint.
+    //
+    // For example, for a CE domain size `n=8`, `slice_indices_precomputes = [0, 8, 12, 14]`, such
+    // that transition constraint `idx` owns the range:
+    // idx=0: [0, 8)
+    // idx=1: [8, 12)
+    // idx=2: [12, 14)
+    slice_indices_precomputes: Vec<usize>,
+}
+
+impl<E: FieldElement> LagrangeKernelTransitionConstraintsDivisor<E> {
+    pub fn new(
+        num_lagrange_transition_constraints: usize,
+        domain: &StarkDomain<E::BaseField>,
+    ) -> Self {
+        let divisor_evals_inv = {
+            let divisor_evaluator = TransitionDivisorEvaluator::<E>::new(
+                num_lagrange_transition_constraints,
+                domain.offset(),
+            );
+
+            // The number of divisor evaluations is
+            // `ce_domain_size + ce_domain_size/2 + ce_domain_size/4 + ... + ce_domain_size/(log(ce_domain_size)-1)`,
+            // which is slightly smaller than `ce_domain_size * 2`
+            let mut divisor_evals: Vec<E> = Vec::with_capacity(domain.ce_domain_size() * 2);
```

Comment:
Probably won't change things much, but we frequently use `uninit_vector()` function to skip memory initialization when the whole vector will be overwritten.

Reply:
`Vec::with_capacity()` doesn't initialize the memory, as discussed in the [Guarantees section of the docs](https://doc.rust-lang.org/std/vec/struct.Vec.html#guarantees) (see the diagram). So unless I'm missing something, the following 2 approaches generate the same code:
- construct vector with `Vec::with_capacity()`, and `Vec::push()` elements one by one
- construct vec with `uninit_vector()`, and write elements one by one with `v[i] = ...;`

Since `Vec::with_capacity()` is canonical, and doesn't involve exposing uninitialize memory, I think it should be preferred over `uninit_vector()`. Unless I'm missing something?

Reply:
In the testing I did a while back (so, things may have changed since then) using un-initialized vector and then setting values via `v[I] = x` was noticeably faster than doing `Vec::with_capacity() ` and then using `.push()` method. Was not a huge difference, but iirc it was somewhere between 10% and 20%.

Reply:
Oh yes you're right, I just benchmarked it, and indeed I using `uninit_vector()` was 40% faster! I will change the code to use it.

Reply:
Actually it makes the code much more complex, since

1. we need to compute the exact size of the vector, which is

```rust
let mut divisor_evals = unsafe {
    uninit_vector(
        domain.ce_domain_size() * 2
            - (domain.ce_domain_size()
                / 2_usize.pow(num_lagrange_transition_constraints as u32 - 1)),
    )
};
```

2. the index to which we need to write to is a complex function of `trans_constraint_idx` and `step`, so we'd be better off introducing an `idx` variable which we increment after each write (which makes the code that much harder to read).

I tried it, ran the benchmarks, and it didn't change the performance. So I would just leave it as is.